docs/proposals/llm-hyperparameters-tuning-api.md

```diff
@@ -0,0 +1,135 @@
+
+# Define hyperparameters and search space
+parameters = {
+	"learning_rate": katib.search.double(min=[], max=[]),
+	"batch_size": katib.search.int(min=[], max=[]),
+	"num_epoch": katib.search.int(min=[], max=[]),
+	"weight_decay": katib.search.double(min=[], max=[])
+	...
+}
+
+# Create experiment for tuning
+exp_name = "llm-tune-test"
+katib_client.tune(
+	name=exp_name,
+	model=model_args,
+	dataset=dataset_args,
+	parameters=parameters, 
+	objective_metric_name: str = None,
+	additional_metric_names: List[str] = [],
+	objective_type: str = "maximize", 
+	objective_goal: float = None,
+	algorithm_name: str = "random",
+	resources_per_trial: Union[dict, client.V1ResourceRequirements, None] = None, 
+	parallel_trial_count: int = None,
+	max_trial_count: int = None,
+	max_failed_trail_count: int = None
+)
+
+# Monitor experiment progress
+katib_client.get_experiment_conditions(exp_name)
+katib_client.get_success_trail_details(exp_name)
+
+# Get the best hyperparameters
+best_hps = katib_client.get_optimal_hyperparameters(exp_name)
+```
+
+### Example
+
+```python
+import kubeflow.storage_initializer.hugging_face
+from kubeflow import katib
+
+# Import model and dataset
+model_args = HuggingFaceModelParams(
+        model_uri="hf://google-bert/bert-base-cased",
+        transformer_type=transformers.AutoModelForSequenceClassification,
+)
+
+dataset_args = HfDatasetParams(
+        repo_id="yelp_review_full",
+        split="train[:3000]",
+)
+
+# Define hyperparameters and search space
+parameters = {
+	"learning_rate": katib.search.double(min=1e-05, max=5e-05),
+	"batch_size": katib.search.int(min=8, max=64),
+	"num_epoch": katib.search.int(min=1, max=10),
+	"weight_decay": katib.search.double(min=0.0, max=1.0)
+	...
+}
+
+# Create experiment for tuning
+exp_name = "llm-tune-test"
+katib_client.tune(
+	name=exp_name,
+	model=model_args, 
+	dataset=dataset_args, 
+	parameters=parameters, 
+	objective_metric_name="accuracy", 
+	objective_type="maximize", 
+	algorithm_name="random",
+	resources_per_trial={"gpu": "2"},
+	parallel_trial_count=2,
+	max_trial_count=50
+)
+
+# Get the best hyperparameters
+best_hps = katib_client.get_optimal_hyperparameters(exp_name)
+```
+
+## Implementation
+
+We will leverage [existing assets](https://github.com/kubeflow/training-operator/tree/687f0c9d2f5cf5dcc97dec87c869ec7f1309d07c/sdk/python/kubeflow/storage_initializer) from the Training Operator to enable the import of pre-trained models and datasets from external platforms (e.g., HuggingFace, S3, etc.). This involves utilizing a storage initializer (PVC) to download pre-trained models and datasets, then sharing this volume with other containers responsible for computation. Additionally, we will define the init container as well as containers for the master and workers in the trail specification for PyTorchJob, similar to the implementation of the train API in Training Operator. This allows us to distribute trials to different workers effectively.
```

Comment:
I guess, initially we will support only PyTorchJob as Trial, right ?

Reply:
Do you have any suggestions on this? I reviewed the `train` API and it appears to support both PyTorchJob and TFJob. However, I am not entirely certain about this.

Reply:
We don't support TFJob in `train` API today, it is only for PyTorch distributed.